src/category_theory/is_connected.lean

```diff
@@ -157,6 +157,19 @@ begin
   rw [w j, w j'],
 end)
 
+/--
+Another induction principle for `is_preconnected J`:
+given a type family `Z : J → Sort*` and
+a rule for transporting in *both* directions along a morphism in `J`,
+we can transport an `x : Z j₀` to a point in `Z j` for any `j`.
+-/
+def is_preconnected_induction [is_preconnected J] (Z : J → Sort*)
+  (h₁ : Π {j₁ j₂ : J} (f : j₁ ⟶ j₂), Z j₁ → Z j₂)
+  (h₂ : Π {j₁ j₂ : J} (f : j₁ ⟶ j₂), Z j₂ → Z j₁)
+  {j₀ : J} (x : Z j₀) (j : J) : Z j :=
+nonempty.some (induct_on_objects {j | nonempty (Z j)} ⟨x⟩
+  (λ j₁ j₂ f, ⟨by { rintro ⟨y⟩, exact ⟨h₁ f y⟩, }, by { rintro ⟨y⟩, exact ⟨h₂ f y⟩, }⟩) j : _)
```

Comment:
Do you want to include some properties about the generated function? For instance, it's not immediate to me that the function you get is uniquely defined - might you need some consistency properties about the `h`s? I could be missing something

Reply:
Since the function constructed this way has no particular properties at all, I've changed the return type to `nonempty (Z j)`.

Over in `cofinal.lean`, I've reduced the induction principle there to one valued in `Prop`, even though `Sort *` works just fine, to keep things simple.